runtime/src/main/java/org/corfudb/runtime/view/stream/SimpleView.java

```diff
@@ -0,0 +1,358 @@
+                        return -1L;
+                    }
+                }
+                // Request a new token, informing the sequencer we were
+                // overwritten.
+                tokenResponse = runtime.getSequencerView()
+                        .nextToken(Collections.singleton(id),
+                                1);
+            } catch (StaleTokenException te) {
+                log.trace("Token grew stale occurred at {}", tokenResponse);
+                if (deacquisitionCallback != null && !deacquisitionCallback.apply(tokenResponse)) {
+                    log.debug("Deacquisition requested abort");
+                    return -1L;
+                }
+                // Request a new token, informing the sequencer we were
+                // overwritten.
+                tokenResponse = runtime.getSequencerView()
+                        .nextToken(Collections.singleton(id),
+                                1);
+
+            }
+        }
+
+        log.error("append[{}]: failed after {} retries, write size {} bytes",
+                tokenResponse.getTokenValue(),
+                runtime.getParameters().getWriteRetry(),
+                ILogData.getSerializedSize(object));
+        throw new AppendException();
+    }
+
+    @Override
+    public ILogData previous() {
+        long address = sas.previous();
+        if (address == Address.NON_ADDRESS) {
+            return null;
+        } else {
+            return read(address);
+        }
+    }
+
+    @Override
+    public ILogData current() {
+        if (sas.getCP() == Address.NON_ADDRESS) {
+            return null;
+        } else {
+            return read(sas.getCP());
+        }
+    }
+
+    void syncAddressSpace(long globalAddress) {
+        if (globalAddress <= globalMax) {
+            return;
+        } else {
+            long oldTail = sas.getMax();
+            Token token = runtime.getSequencerView()
+                    .nextToken(Collections.singleton(id), 0).getToken();
+            long newTail = token.getTokenValue();
+
+            List<Long> addressesToAdd = new ArrayList<>(100);
+
+            while (newTail > oldTail) {
+                ILogData d = read(newTail);
+                if (d.containsStream(id)) {
+                    addressesToAdd.add(newTail);
+                }
+                // What happens when a hole is encountered ?
+                newTail = d.getBackpointer(id);
+            }
+
+            // update global max and add the new tail addresses
+            // need to consider the global tail too !!!!!!!
+            //globalMax = Math.max(globalMax, token.getGlobalTail());
+            //System.out.println("---> globalAddress " + globalAddress);
+            globalMax = Math.max(globalMax, globalAddress);
+            List<Long> revList = Lists.reverse(addressesToAdd);
+            sas.addAddresses(revList);
+        }
+    }
+
+    @Override
+    public ILogData next() {
```

Comment:
if the syncAddressSpace method is part of the inner class you could write this simpler
if(sas.hasNext())
    return read(sas.next())
else
    return null;